agent/consul/state/catalog.go

```diff
@@ -2820,3 +2908,348 @@ func checkProtocolMatch(tx ReadTxn, ws memdb.WatchSet, svc *structs.GatewayServi
+) (uint64, *structs.ServiceTopology, error) {
+
+	var (
+		maxIdx uint64
+		sn     = structs.NewServiceName(service, entMeta)
+	)
+	idx, upstreamNames, err := s.UpstreamsForService(ws, dc, sn)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to get upstreams for %q: %v", sn.String(), err)
+	}
+	if idx > maxIdx {
+		maxIdx = idx
+	}
+
+	var upstreams structs.CheckServiceNodes
+	for _, u := range upstreamNames {
+		// Collect both typical and connect endpoints, this allows aggregating check statuses across both
+		idx, csn, err := s.CheckServiceNodes(ws, u.Name, &u.EnterpriseMeta)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to get upstream nodes for %q: %v", sn.String(), err)
+		}
+		if idx > maxIdx {
+			maxIdx = idx
+		}
+		upstreams = append(upstreams, csn...)
+
+		idx, csn, err = s.CheckConnectServiceNodes(ws, u.Name, &u.EnterpriseMeta)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to get upstream connect nodes for %q: %v", sn.String(), err)
+		}
+		if idx > maxIdx {
+			maxIdx = idx
+		}
+		upstreams = append(upstreams, csn...)
+	}
+
+	idx, downstreamNames, err := s.DownstreamsForService(ws, dc, sn)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to get downstreams for %q: %v", sn.String(), err)
+	}
+	if idx > maxIdx {
+		maxIdx = idx
+	}
+
+	var downstreams structs.CheckServiceNodes
+	for _, u := range downstreamNames {
+		// Collect both typical and connect endpoints, this allows aggregating check statuses across both
+		idx, csn, err := s.CheckServiceNodes(ws, u.Name, &u.EnterpriseMeta)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to get downstream nodes for %q: %v", sn.String(), err)
+		}
+		if idx > maxIdx {
+			maxIdx = idx
+		}
+		downstreams = append(downstreams, csn...)
+
+		idx, csn, err = s.CheckConnectServiceNodes(ws, u.Name, &u.EnterpriseMeta)
+		if err != nil {
+			return 0, nil, fmt.Errorf("failed to get downstream connect nodes for %q: %v", sn.String(), err)
+		}
+		if idx > maxIdx {
+			maxIdx = idx
+		}
+		downstreams = append(downstreams, csn...)
+	}
+
+	resp := &structs.ServiceTopology{
+		Upstreams:   upstreams,
+		Downstreams: downstreams,
+	}
+	return 0, resp, nil
+}
+
+// UpstreamsForService will find all upstream services that the input could route traffic to.
+// There are two factors at play. Upstreams defined in a proxy registration, and the discovery chain for those upstreams.
+// TODO (freddy): Account for ingress gateways
+func (s *Store) UpstreamsForService(ws memdb.WatchSet, dc string, sn structs.ServiceName) (uint64, []structs.ServiceName, error) {
+	tx := s.db.ReadTxn()
+	defer tx.Abort()
+
+	idx, upstreams, err := upstreamsFromRegistration(ws, tx, sn)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to get registration upstreams for %q: %v", sn.String(), err)
+	}
+
+	var maxIdx uint64
+	if idx > maxIdx {
+		maxIdx = idx
+	}
+
+	var (
+		resp []structs.ServiceName
+		seen = make(map[structs.ServiceName]bool)
+	)
+	for _, u := range upstreams {
```

Comment:
The upstream might have the `datacenter` in it already if someone overrode it, which you might be able to use here if you split the `dc` arg on `discoveryChainTargets` to have two (the evaluate-in and use-in variety).

Reply:
Yep, will thread that through in a followup PR with multi-DC upstreams

Reply:
Yep, done.